plaso/filters/parser_filter.py

```diff
@@ -0,0 +1,227 @@
+    """Creates an expression string of the excluded and included parsers.
+
+    Args:
+      excludes (dict[str, set[str]]): excluded parsers and plugins.
+      includes (dict[str, set[str]]): included parsers and plugins.
+
+    Returns:
+      str: a parser filter expression or None to represent all parsers and
+           plugins.
+
+    Raises:
+      RuntimeError: if a specific plugin is excluded but no corresponding parser
+          is included or if exclude and include parser filters overlap.
+    """
+    excluded_parsers_and_plugins = self._GetParserAndPluginsList(excludes)
+    included_parsers_and_plugins = self._GetParserAndPluginsList(includes)
+
+    # Note that below set comprehension is used to determine the set of
+    # excluded and included parsers.
+    excluded_parsers = {
+        name for name, plugins in excludes.items() if plugins != set(['*'])}
+    included_parsers = {
+        name for name, plugins in includes.items() if '*' in plugins}
+    missing_parser_filters = excluded_parsers.difference(included_parsers)
+    if missing_parser_filters:
+      missing_parser_filters = self._GetParserAndPluginsList({
+          name:  excludes[name] for name in missing_parser_filters})
+      raise RuntimeError((
+          'Parser filters: {0:s} defined to be excluded but no corresponding '
+          'include expression').format(','.join(missing_parser_filters)))
+
+    overlapping_parser_filters = set(included_parsers_and_plugins).intersection(
+        set(excluded_parsers_and_plugins))
+    if overlapping_parser_filters:
+      raise RuntimeError((
+          'Parser filters: {0:s} defined to be both included and '
+          'excludes').format(','.join(overlapping_parser_filters)))
+
+    parser_filters = [
+        '!{0:s}'.format(parser_filter)
+        for parser_filter in excluded_parsers_and_plugins]
+    parser_filters.extend(included_parsers_and_plugins)
+    return ','.join(parser_filters)
+
+  def _ExpandPreset(self, presets_manager, preset_name, parsers_and_plugins):
+    """Expands a preset in the parsers and plugins.
+
+    This functions replaces the preset in parsers_and_plugins with the parser
+    and plugin or preset names defined by the preset.
+
+    Args:
+      presets_manager (ParserPresetsManager): a parser preset manager, that
+          is used to resolve which parsers and/or plugins are defined by
+          presets.
+      preset_name (str): name of the peset to expand.
+      parsers_and_plugins (dict[str, set[str]]): parsers, plugins and presets.
+
+    Raises:
+      RuntimeError: if the plugins list of a preset in parsers_and_plugins,
+          contains plugin definitions other than '*'.
+    """
+    if preset_name not in parsers_and_plugins:
+      return
+
+    plugins = parsers_and_plugins.get(preset_name, None)
+    if plugins != set(['*']):
+      print(plugins)
+      raise RuntimeError('Unsupported preset: {0:s} with plugins.'.format(
```

Comment:
Seeing this is already a reworded version. What would help if you provide suggestions which wording that you do understand, or asking clarifying question. Instead of creating yet another round trip to get clarified what part is not clear to you.

This is an internal error. It means that the preset in the parsers_and_plugins dict is used as a preset and defines plugins, which should not be the case.